br/pkg/stream/rewrite_meta_rawkv.go

```diff
@@ -0,0 +1,253 @@
+		NewSchemas: NewSchemas,
+		RewriteTS:  restoreTS,
+	}
+}
+
+func (sr *SchemasReplace) rewriteKeyForDB(key []byte, cf string) ([]byte, error) {
+	rawMetaKey, err := ParseTxnMetaKeyFrom(key)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	dbID, err := meta.ParseDBKey(rawMetaKey.Field)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	dbName := sr.OldDBs[dbID].Name.String()
+	newDbID := sr.NewSchemas[dbName].DbInfo.ID
+
+	rawMetaKey.UpdateField(meta.DBkey(newDbID))
+	return rawMetaKey.EncodeMetaKey(), nil
+}
+
+func (sr *SchemasReplace) rewriteValueForDB(value []byte, cf string) ([]byte, error) {
+	var (
+		newValue []byte
+		dbInfo   model.DBInfo
+		err      error
+	)
+
+	if cf == "default" {
+		if err = json.Unmarshal(value, &dbInfo); err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		dbName := dbInfo.Name.String()
+		newDbID := sr.NewSchemas[dbName].DbInfo.ID
+		dbInfo.ID = newDbID
+		if newValue, err = json.Marshal(&dbInfo); err != nil {
+			return nil, err
+		}
+	} else if cf == "write" {
+		rawWriteCFValue := new(RawWriteCFValue)
+		if err := rawWriteCFValue.ParseFrom(value); err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		if !rawWriteCFValue.HasShortValue() {
+			return value, nil
+		}
+
+		shortValue := rawWriteCFValue.GetShortValue()
+		if err := json.Unmarshal(shortValue, &dbInfo); err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		dbName := dbInfo.Name.String()
+		newDbID := sr.NewSchemas[dbName].DbInfo.ID
+		dbInfo.ID = newDbID
+		if shortValue, err = json.Marshal(&dbInfo); err != nil {
+			return nil, err
+		}
+		rawWriteCFValue.UpdateShortValue(shortValue)
+		newValue = rawWriteCFValue.EncodeTo()
+	} else {
+		panic(fmt.Sprintf("not support cf:%s", cf))
+	}
+
+	return newValue, nil
+}
+
+func (sr *SchemasReplace) rewriteKVEntryForDB(e *kv.Entry, cf string) (*kv.Entry, error) {
+	newKey, err := sr.rewriteKeyForDB(e.Key, cf)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	newValue, err := sr.rewriteValueForDB(e.Value, cf)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &kv.Entry{Key: newKey, Value: newValue}, nil
+}
+
+func (sr *SchemasReplace) rewriteKeyForTable(key []byte, cf string) ([]byte, string, error) {
+	rawMetakey, err := ParseTxnMetaKeyFrom(key)
+	if err != nil {
+		return nil, "", errors.Trace(err)
+	}
+
+	dbID, err := meta.ParseDBKey(rawMetakey.Key)
+	if err != nil {
+		return nil, "", errors.Trace(err)
+	}
+
+	tableID, err := meta.ParseTableKey(rawMetakey.Field)
+	if err != nil {
+		log.Info("parse table key failed", zap.ByteString("field", rawMetakey.Field))
+		return nil, "", errors.Trace(err)
+	}
+
+	dbName := sr.OldDBs[dbID].Name.String()
+	tableName := sr.OldTables[tableID].Name.String()
+	newDbID := sr.NewSchemas[dbName].DbInfo.ID
+	newTableID := sr.NewSchemas[dbName].Tables[tableName].ID
+
+	rawMetakey.UpdateKey(meta.DBkey(newDbID))
+	rawMetakey.UpdateField(meta.TableKey(newTableID))
+	if cf == "write" {
+		rawMetakey.UpdateTS(sr.RewriteTS)
```

Comment:
I'm afraid some type of DDLs (like `rename table`) may write the same key twice or more times. If so, once we rewrite all keys to the same TS, we may lost the order of those keys.